test patient MRN finder

dantohe opened the pull request:
please review test patient MRN finder

dantohe committed:
moved the patient finder here

dantohe committed:
added test folder

dantohe committed:
added test files

dantohe committed:
csmetic changes in the finder

dantohe committed:
cosmetic changes

dantohe committed:
cosmetic changes

Xodarap commented:
To run: `bundle exec ruby findPatientsIDs.rb test/T63-22.zip`
if it doesn't work check if this site is running http://54.152.244.137/patients

dantohe commented:
This utility finds MRNs (unique medical record numbers) for a set of QRDA1 files. 
These QRDA1 files are test files created by Cypress for certification - they are patient records to be evaluated against measures. 
This utility:
1. takes a zip file containing the test records (QRDA1 files) and  parses them in healt-data-standards records (creates set#1).
2. grabs the set of master records from Cypress mongo database and puts them in healt-data-standards records (creates set#2) - (master records are used internally by Cypress to create test records)
3. does deep comparison at the record section code level (like encounter, medications etc) between  set#1 and set#2 - if match exists (a test record codes are contained in the master record) then the master record is considered as candidate 
4. addresses the issue that for some test records multiple candidates are found by manual annotation

Xodarap commented on lib/patientFinder/findPatientsIDs.rb:
```diff
@@ -0,0 +1,203 @@
+			#simple conditions that will reduce the record pool
+			if(masterRecord.gender==testRecord.gender && masterRecord.ethnicity[:code]==testRecord.ethnicity[:code] && masterRecord.race[:code] == testRecord.race[:code])
+				#go inside the sections and collect codes and find if the codes from the test are ALL contained in the the same section of the master 
+				allergiesMatch=Utils::matchCodes(testRecord.allergies,masterRecord.allergies)
```
These next few lines can be replaced with:

```
sections = [:allergies, :care_goals,…]
master_is_match = sections.all? {| section | Utils::matchCodes(testRecord[section],masterRecord[section])}
```

dantohe replied:
couldn't get this to work  - I tried myself to figure out a more elegant way to deal with the sections but no success so far - maybe we can take a look together

Xodarap replied:
Weird, yeah let's take a look tomorrow if you have time

Xodarap commented on lib/patientFinder/findPatientsIDs.rb:
```diff
@@ -0,0 +1,203 @@
+			createCodeHash(thisEntry, masterCodesHash)
+		end
+
+		if (testCodesHash.to_a - masterCodesHash.to_a).empty?
```
This is more of a preference than anything else, but usually people will just return the value of `(testCodesHash.to_a - masterCodesHash.to_a).empty?` rather than creating a dummy variable

dantohe replied:
indeed

Xodarap commented on lib/patientFinder/findPatientsIDs.rb:
```diff
@@ -0,0 +1,203 @@
+	#a utility that iterates through a set of master records to find possible matches (candidates) for a test cord
+	def self.iterateThroughMasterRecordsAndFindCandidates(masterRecords,testRecord)
+		candidates=Array.new 
+		for masterRecord in masterRecords
```
`masterRecords.each`is considered more idiomatic than `for`, see http://graysoftinc.com/early-steps/the-evils-of-the-for-loop

dantohe replied:
acknowledged - interesting - I found this on https://github.com/styleguide/ruby
Never use for, unless you know exactly why. Most of the time iterators should be used instead. for is implemented in terms of each (so you're adding a level of indirection), but with a twist - for doesn't introduce a new scope (unlike each) and variables defined in its block will be visible outside it.

Xodarap commented on lib/patientFinder/findPatientsIDs.rb:
```diff
@@ -0,0 +1,203 @@
+require 'health-data-standards'
+
+#a simple patient from QRDA - contains the source files name and the parsed record associated
+class PatientFromQRDA
```
In general, I think one class per file is a good idea, even if the classes are really short like this

dantohe replied:
Agreed - my assumption was that Ruby is rather liberal with this - I need to read something on Ruby coding style

Xodarap commented on lib/patientFinder/findPatientsIDs.rb:
```diff
@@ -0,0 +1,203 @@
+
+	#extracts the ids from a set of records 
+	def self.collectMedicalRecordNumbers(candidates)
+		ids=Array.new
```
This function can be replaced with `candidates.map(&:medical_record_number)`. In general you don't have to declare variables in ruby before using them (even implicitly like by setting them to be `Array.new`)

dantohe replied:
indeed it works .... it was a bit magic before I read about it ....  ( referring to candidates.map(&:medical_record_number)) - I get it regarding the "In general you don't have to declare variables in ruby before using them"

Xodarap commented on lib/patientFinder/mongoid.yml:
```diff
@@ -0,0 +1,34 @@
+development:
+  sessions:
+    default:
+      database: patid
```
We are still figuring out how we want to implement multiple databases ourselves, but my understanding is that usually applications to not look across databases (so instead this would usually be implemented by creating a separate collection in the same database)

dantohe replied:
that's fine - even if we save the test records on the same collection (Cypress db records collection) it shouldn't be a a problem as the master records will not be affected in any way

dantohe committed:
some of the changes proposed in the code review

dantohe committed:
moved classes in separate files, refactored code

dantohe committed:
test infrastructure

dantohe committed:
test cases for MRNfinder

dantohe committed:
added a few more tests around MRNfinder, moved get_MRNs into utils

dantohe committed:
Changed the test directory structure:
-added a unit directory
-moved mRNfinder_test.rb in the unit directory
-changed the Rakefile to reflect this change

dantohe committed:
Add unit tests for PatientFromQRDA
-created a QRDA1 example that will be imported by the setup for unitTest patientFromQRDA_test
-created 2 unitTests in patientFromQRDA_test

dantohe committed:
Fix the multiple mongoDb instances:

dantohe committed:
Adresses the code  ugliness regarding ambiguous patinets:
- Moved manual annotations into a constant hash
- Created a new method that resolves them

dantohe committed:
Refactor of utils.rb

Xodarap commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,165 @@
+    class Utils
+        
+        #some of records are created ambiguous on purpose (fr example some of them contain only encounters) - in order to disambiguate human intervention is needed - the following hash resolves the ambiguos cases
+        AMBIGUOUS_PATIENTS = {["ee85c89f24946e2ddca12c6edc5181dc", "eb6af018e11320f5c163624beb83767d"] => "ee85c89f24946e2ddca12c6edc5181dc"}            
```
Just a formatting comment: usually hashes are created all within one set of curly braces, rather than doing one value per line like you have

dantohe replied:
I know  - I initially have it in one single block between {} - after that I changed to multiple line declaration to make it easier to read - but if the general convention is to have everything between {} I am conforming to it. Nathan mention below are good - unfortunately I couldn't replace {} w/o getting syntax errors - do you have any example for hash build using do...end

Xodarap commented:
Usually tests are placed in the highest level test folder, rather than having separate test folders and rake files inside subdirectories.

Xodarap commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,165 @@
+                    # }
+                    # sections_with_data=get_non_empty_sections_for_record(sections, test_record)
+                    
+                    # master_is_match = sections_with_data.all? {|section| Utils::codes_match?(test_record[section],master_record[section])}
```
If I recall correctly, the problem is that treating the object like a hash will return nil instead of an empty array if the section doesn't exist. One way you could handle this is to do `test_record.send section` instead of `test_record[section]` and similarly for master_record.

nbouscal replied:
A different change that also appears to fix this is to replace `thisEntry.codes` with `thisEntry[:codes]` on line 141.

Edit: Oh, no, this works on the tests but not if I run the whole thing. You could still make it work there by adding `entries ||= []` before the `each` block.

nbouscal commented:
I have a somewhat different testing philosophy from what I'm seeing here. As an example for discussion, here's one test that I would write differently:

```
def test_get_hash_out_of_entries
    assert_equal( {"99201"=>"CPT"} , PatientFinder::Utils.get_hash_out_of_entries(@test_patient_from_qrda.record.encounters))
end
```

I would instead write something like this:

```
def test_get_hash_out_of_entries
    entries = [{codes: {set1: [:value1, :value2], set2: [:value3]}},
               {codes: {set3: [:value4]}}]
    code_hash = {value1: :set1, value2: :set1, value3: :set2, value4: :set3}
    result = PatientFinder::Utils.get_hash_out_of_entries(entries)
    assert_equal(result, code_hash)
end
```

There are a few different things at play here. First, since the patient data being used in the test currently only has a single code, the test doesn't actually guarantee that the method is doing what you want it to.  For example, if I change the method to the following, your test still passes:

```
def self.get_hash_out_of_entries(entries)
    code_hash = Hash.new
    thisEntry = entries.first
    thisEntry[:codes].each_pair do |code_set, coded_values|
        coded_value = coded_values.first
        code_hash = {coded_value => code_set}
    end
    code_hash
end
```

I think there's a fair argument that tests should not go too far in being defensive against malicious changes, but I think this one is a little bit too easy to trick. If you aren't familiar with it, [mutation testing](https://en.wikipedia.org/wiki/Mutation_testing) is a pretty cool thing in this space.

Second, your current test does little to help me understand what the method is doing. The `get_hash_out_of_entries` function is really a domain-agnostic data structure transformation. The only domain information that it cares about is the key name `codes`. If I want to understand the transformation from the test, though, I have to figure out what data structure that patient is going to have in their `record.encounters` attribute. I can do that by using a `binding.pry` to look at the actual patient during the test, but I'd rather not have to. I can also try to look at the source XML file from which the patient is generated, but the string `CPT` doesn't actually appear in that file. Tests are no replacement for documentation, but clear tests _can_ make understanding the code a lot easier, by giving you a good example of the type of thing a given method should be doing.

Third, and most importantly, your current test really isn't a unit test. You're testing a domain-agnostic function, but you're bringing in a ton of domain stuff to do so. That means time overhead as all of the stuff in the `setup` block runs, and it also means that the test can fail even if the `get_hash_out_of_entries` function remains exactly the same. If any of the logic going into creating that patient fails, there's some chance this test would fail even though it shouldn't.

I'm also not actually sure this function should be part of this class's API. It's only used by a single method, `codes_match?`, and it seems to be pretty implementation detail-ish. I would consider making it private, or possibly even just a lambda inside of `codes_match?`. In either of those cases, you would end up not testing it at all, instead just testing the `codes_match?` function. That seems pretty much fine, since `codes_match?` has very little logic aside from this. In either case, my comments about `test_codes_match?` would be similar to the ones I've made here.

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,165 @@
+                    insurance_providersMatch = codes_match?(test_record.insurance_providers,master_record.insurance_providers)
+    
+                    #create candidates based on matching codes - if all the above match then consider this master as candidate
+                    if(allergiesMatch && care_goalsMatch && conditionsMatch && encountersMatch &&  immunizationsMatch && medical_equipmentMatch &&  medicationsMatch && proceduresMatch &&  insurance_providersMatch)
```
If you were to continue doing it this way, a slightly nicer way of doing this would be:

```
matches = [allergiesMatch, care_goalsMatch, conditionsMatch, encountersMatch, immunizationsMatch, medical_equipmentMatch, medicationsMatch, proceduresMatch, insurance_providersMatch]
if (matches.all?)
    candidates << master_record
end
```

It could then be made even nicer by replacing each assignment with a shovel onto the array, i.e.

```
matches = []
matches << codes_match?(test_record.allergies, master_record.allergies)
matches << codes_match?(test_record.care_goals, master_record.care_goals)
...
matches << codes_match?(test_record.insurance_providers, master_record.insurance_providers)

if (matches.all?)
    candidates << master_record
end
```

Or, similarly, you could just have a boolean and use `&&=` at each line.

I still much prefer your original way with one of the fixes mentioned above, though.

dantohe replied:
well - test_record.send(section), master_record.send(section) did the trick - really happy as I agonized on this - thanks!

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,165 @@
+            master_records.each { |master_record|
+            
+                #simple conditions that will reduce the record pool
+                if(master_record.gender==test_record.gender && master_record.ethnicity[:code]==test_record.ethnicity[:code] && master_record.race[:code] == test_record.race[:code])
```
At minimum, I would prefer to have this broken up on multiple lines. We don't enforce an 80-character limit currently (partly because we inherited a lot of code that would violate it), but it's still a nice thing to target. So, e.g.

```
if (master_record.gender == test_record.gender &&
    master_record.ethnicity[:code] == test_record.ethnicity[:code] &&
    master_record.race[:code] == test_record.race[:code])
```

At this indentation level the second line of that still hits 83 (which is one reason I usually prefer 2-space indentation to 4-space).

I would also consider a more significant refactor, possibly something like this:

```
demo = Struct.new(:gender, :ethnicity, :race)
get_demo = -> (obj) do
  demo.new(obj.gender, obj.ethnicity[:code], obj.race[:code])
end
if (get_demo.(master_record) == get_demo.(test_record))
```

Xodarap replied:
I personally would not go as far as the second re-factoring, although I agree that having such bloated if statements is ugly. I agree with the line breaking suggestion.

nbouscal replied:
Yeah that refactoring is most likely me trying to write Haskell in Ruby :P

dantohe replied:
nevertheless,  interesting construct ... - I will settle for breaking the lines ....

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,165 @@
+                test_to_MRN_hash[test_record.fileName]=candidate_MRNs.first
+            else
+                test_to_MRN_hash[test_record.fileName]=get_MRN_for_ambiguous(candidate_MRNs)
+            end
```
In Ruby, `if`s are expressions, which means you can do this instead:

```
test_to_MRN_hash[test_record.fileName] =
  if (candidate_MRNs.length == 1)
    candidate_MRNs.first
  else
    get_MRN_for_ambiguous(candidate_MRNs)
  end
```

dantohe replied:
yes - I keep forgetting about this ....

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,165 @@
+        end
+    
+        #resolves the candidates to the test files (one test file - one candidate - when multiple candidates are found manual anotation is used)
+        def self.get_MRN_for_tests(candidate_MRNs, test_record, test_to_MRN_hash)
```
Rather than passing a hash as an argument and then mutating it, I think it's more straightforward/idiomatic for this method to simply return the result, and have the hash assignment happen in `get_MRNs`.

dantohe replied:
Indeed - what you describes makes it better

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,165 @@
+        #         end
+        #     }
+        #     sections_with_data
+        # end
```
This method can be implemented this way:

```
def self.get_non_empty_sections_for_record(sections, record)
  sections.select { |section| record[section].present? }
end
```

That being the case, I would probably just inline it where it is being called in `get_candidates_for_test`.

This implementation is not completely identical, because `.present?` is the inverse of `.blank?`, which is [slightly different](https://stackoverflow.com/questions/885414/a-concise-explanation-of-nil-v-empty-v-blank-in-ruby-on-rails) from `.nil? || .empty?`, but those differences should not be relevant in this case.

dantohe replied:
nice - .select and .map still stuff to learn for me ....

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,165 @@
+                
+                #resolve cadidates to unique MRNs - with other words resolve the candidates
+                get_MRN_for_tests(candidates.map(&:medical_record_number), test_patient, tests_to_MRNs)
+            }
```
Very minor style thing: `{...}` is equivalent to `do...end`, but it's preferable to use `do...end` if the block spans multiple lines (and likewise, to use `{...}` if it doesn't).

dantohe replied:
good - unfortunately I couldn't replace {} w/o getting syntax errors - do you have any example (tried to find one but no success) for hash build using do...end?

nbouscal replied:
Oh, sorry, the thing I said was not strictly true. They're equivalent for blocks, but are not equivalent in general. For hash literals, you can only use braces (and it's fine to do so over multiple lines). I was only suggesting using `do...end` for this `each` block, and other similar things.

dantohe committed:
Add two zip test fixtures:
- this should address the issue of testing taking too long due to too many patients being processed for testing purposes only

dantohe committed:
Fixed the problematic archives commited in the previos step

dantohe committed:
Fix for Unit Tests taking too long:
- only a limited set of QRDA patients are used
- a separate set of tests with comprehensive coverage is in mRNfinder_masive.rb - the intention is to run this only as per need basis
- with this change the whole test suite runs in less than 4 seconds on the cloud9 env.

dantohe committed:
change AMBIGUOUS_PATIENTS declaration

dantohe committed:
Change for get_MRN_for_tests as per code review recom

dantohe committed:
Adress the ugly code from get_candidates_for_test:
- sections_with_data is done now as per code review suggestion
- master_is_match is also collected as per code review suggestion

dantohe commented:
@Xodarap: regarding "Usually tests are placed in the highest level test folder, rather than having separate test folders and rake files inside subdirectories." - I understand - the thing is that here I had to shortcut a bit the normal Rails infrastructure and my tests are placed in the wrong folder (I guess they should rather be in Cypress/test. If I well understand you are also saying that they should not be placed in a folder called unit but rather in the test directly - OK - I looked in the Cypress tests and they had a unit directory.

@nbouscal - I accept your last point regrading test_codes_match? & get_hash_out_of_entries  (they do not bring value to the API) - so made them private and removed the 2 test cases around them

dantohe committed:
Adresses Test Cases code review observations:
- comment: codes_match? & get_hash_out_of_entries (they do not bring value to the API) - fix: made them private
- changed utils.rb by removing the UT around the two methods

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,142 @@
+                    master_record.race[:code] == test_record.race[:code])
+                    
+                    #go inside the sections and collect codes and find if the codes from the test are ALL contained in the the same section of the master 
+                    sections = [:allergies, :care_goals, :conditions, :encounters, :immunizations, :medical_equipment, :medications, :procedures, :insurance_providers ]
```
Interesting tidbit, Ruby has another way of building an array of symbols:

```
sections = %i(allergies care_goals conditions encounters immunizations medical_equipment medications procedures insurance_providers)
```

It's slightly shorter but also slightly more opaque to people who don't know Ruby well, so I don't have a real preference, just thought I'd let you know about it.

In either style, I think adding some line breaks is worth considering.

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,142 @@
+            
+            candidates=Array.new
+            # candidates
+            master_records.each { |master_record|
```
This whole `each` block would be more idiomatic as a `select` block.

dantohe committed:
Adresses Test Cases code review observations:
- comment: codes_match? & get_hash_out_of_entries (they do not bring value to the API) - fix: made them private
- changed utils.rb by removing the UT around the two methods

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,142 @@
+        def self.get_MRN_for_tests(candidate_MRNs)
+            
+            return candidate_MRNs.first if(candidate_MRNs.length==1)
+            return get_MRN_for_ambiguous(candidate_MRNs) if (candidate_MRNs.length>1)
```
I think this would be more readable this way, but it's just a style thing:

```
if (candidate_MRNs.length > 1)
  get_MRN_for_ambiguous(candidate_MRNs)
else
  candidate_MRNs.first
end
```

This way you only check the length once, and don't need explicit `return`s.

Also in both styles, `get_MRN_for_tests([]) == nil`.

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,142 @@
+                candidates = get_candidates_for_test(master_records, test_patient.record)
+                
+                #resolve cadidates to unique MRNs - with other words resolve the candidates
+                tests_to_MRNs[test_patient.fileName]=get_MRN_for_tests(candidates.map(&:medical_record_number))
```
I would pull `candidates.map(&:medical_record_number)` onto its own line.

nbouscal commented on lib/patientFinder/utils.rb:
```diff
@@ -0,0 +1,142 @@
+                    sections = [:allergies, :care_goals, :conditions, :encounters, :immunizations, :medical_equipment, :medications, :procedures, :insurance_providers ]
+                    sections_with_data = sections.select { |section| test_record[section].present? }
+                    
+                    master_is_match = sections_with_data.all? {|section| codes_match?(test_record.send(section), master_record.send(section))}
```
I would break this onto multiple lines, by using a `do...end` block rather than the single-line version.

dantohe committed:
Added a MRN find  button on products/show

dantohe committed:
MRN finder functionality moved into the rails framework:
- added a button in products/show.html.erb
- added controller for MRNs
- added the model for MRNs
- added view for MRNs
- changes in lib/patientFinder to accomodate the rails integration
- still TODO code cleanup, tests

dantohe committed:
MRN finder functionality - added the missing files from yesterday

dantohe committed:
fixes trailing whitespace issues
